Atomic.md

```diff
@@ -0,0 +1,278 @@
+
+
+    -   The MirrorProvider doesn't use contentID so no issues will occur
+
+    -   VFSForGit will be unable to get the SHA of the object that it
+         needs to download and hydrating this file will fail. Retrying
+         will not help.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   Placeholder does not have flags for virtualization root and
+        empty.
+
+    -   This file will never be expanded as the kext will exit when we
+         do not see the InVirtualizationRoot bit
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Placeholder is marked as in virtualization root but not empty.
+
+        -   This file will never be hydrated as the kext will exit when
+             we do not see the IsEmpty bit
+
+    -   Placeholder is marked as empty but not in virtualization root
+
+        -   This file will never be hydrated as the kext will exit when
+             we do not see the InVirtualizationRoot bit (we'll never
+             reach the IsEmpty test)
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   The default mode should be 755 (unless a user has played with
+         umask, but this hole is out of scope for this design) so
+         hydration of the file should succeed. However, if the file had
+         a mode other than 755, tools that depend on it to be another
+         mode will fail. Recovery would require manual intervention by
+         chmodding this file.
+
+I believe that all of these problems can be addressed by moving to a
+model where we create placeholders in a temporary location, mark them up
+and then rename them into place.
+
+Proposed Solution:
+------------------
+
+This solution creates placeholders in a .temp directory within the
+provider's folder (.gvfs/.mirror), marks them up and then uses rename()
+to move them into the virtualization root.
+
+A sample of this change is implemented here (user mode component only):
+https://github.com/nickgra/VFSForGit/tree/atomic
+
+Currently, the provider registers its virtualization root with the kext
+at clone time. We have code today that will log on files detected
+outside of the virtualization root, which could be helpful for
+diagnosing customer issues.
+
+4974: 218.308) Note: No virtualization root found for file with set
+flag. (vnode path: \'/Users/nickgra/TestRoot/.temp/mov\')
+
+In future, the consequences could be worse than logging, so the idea is
+to wall off the temporary directory so that any activity in it is
+ignored. Changes will be needed on the provider side to register the
+temporary path alongside the kext changes to actually ignore activity in
+registered temp directories in the kext.
+
+Let's enumerate all the failure cases again under this new design:
+
+WritePlaceholderDirectory:
+
+-   mkdir() fails to create the placeholder directory
+
+    -   We can simply retry and have this succeed. Nothing was created
```

Comment:
Do we retry within the same enumeration attempt? Or fail this one and the retry happens the next time the user tries some IO?